modules/self-managed-node-group/main.tf

```
@@ -144,8 +149,20 @@ resource "aws_launch_template" "this" {
     }
   }
 
-  iam_instance_profile {
-    arn = var.create_iam_instance_profile ? aws_iam_instance_profile.this[0].arn : var.iam_instance_profile_arn
+  dynamic "iam_instance_profile" {
+    for_each = !var.create_iam_instance_profile && var.iam_instance_profile_arn != null ? [1] : []
+
+    content {
+      arn = var.iam_instance_profile_arn
+    }
+  }
+
+  dynamic "iam_instance_profile" {
+    for_each = !var.create_iam_instance_profile && var.iam_instance_profile_arn == null && var.iam_role_name != null ? [1] : []
+
+    content {
+      name = var.iam_role_name
+    }
```

This change prevents the instance profiled created w/ `var.create_iam_instance_profile` from being used.

Since we're fixing this here, I think this code is overly complicated so here's my suggestion:
```suggestion
  iam_instance_profile {
    arn = try(aws_iam_instance_profile.this[0].arn, var.iam_instance_profile_arn)
    name = try(aws_iam_instance_profile.this[0].arn, var.iam_instance_profile_arn) == null ? var.iam_role_name : null
  }
```

---

Thanks for the suggestion, it seems reasonable but it doesn't utilize `var.create_iam_instance_profile` like the original code does. I'm wondering if it could lead to some potential issues, or just `try` among two options for both `arn` and `name` is enough to cover all cases.

---

You could just keep the existing code and just add the name line:
```
name = (var.create_iam_instance_profile ? aws_iam_instance_profile.this[0].arn : var.iam_instance_profile_arn) == null ? var.iam_role_name : null

---

Only one of `arn` or `name` can be non-null: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/launch_template#instance-profile

---

Updated accordingly in https://github.com/pinterest/terraform-aws-eks/pull/3/commits/91e58f40aa16c408a07be5e8895ab8a59a88d93f